### ISSUE TITLE
Model Parallelism

### ISSUE THREAD

**Issue by yingfeng** [PaddlePaddle/Paddle#141]
Does paddle have the plan to support model parallelism in distributed environment? Current model parallelism is limited to run on single machine, additionally, it's used only when layers themselves could be parallel where different gpus are assigned with different layers.  A distributed model parallelism means to split certain layers into sublayers running across different machines,  although much larger communication cost is introduced, while under situations where model couldnot be contained on single machine, it's a useful feature.

**Comment by luotao1** [comment]
You can use `parallel_nn` to do model parallelism, please refer to http://www.paddlepaddle.org/doc/ui/cmd_argument/use_case.html#parallel-nn. An example unitest config is https://github.com/baidu/Paddle/blob/master/paddle/trainer/tests/sample_trainer_config_parallel.conf

**Comment by hedaoyuan** [comment]
Implementing an efficient model-parallelism between multiple machines is more difficult. How to split layers into multiple sublayers and the performance of the network are two main problems. At present, more data-parallelism applications, and we more work is on the multi-machine data parallel performance optimization.

**Comment by yingfeng** [comment]
A pipeline based so-called model parallelism might be easier to deliver. For example, a minibatch is sent to device A, the output of device A would be the input of device B, and a new minibatch could be sent to device A after B could process the output of A.

**Comment by emailweixu** [comment]
Currently, there are machines with up to 8 GPU cards commercially available. It is much more efficient to use this kind of machine to train large models. If this kind of machine still cannot hold one single model, the model probably is too big to be useful in real applications.

**Comment by hedaoyuan** [comment]
Currently, paddle can only construct single gradient machine for a neural network model, and one gradient machine can only be set to work on a single GPU. We are thinking of splitting a neural network model into multiple subnetworks, each of which can be constructed as a gradient machine, and each machine can be set to work on different GPUs.

**Comment by backyes** [comment]
@emailweixu 's answer implies most needs for training BIG model can be satisfied with these machine armed with lots of GPUs in PaddlePaddle.  
Indeed, if  some 'poor' user could not get these well armed GPU machines, maybe they really are not going to train their BIG model, even lots of 'poor' armed GPU machines are provided. For these users, distributed model parall. could be preferred, but I believed that the overall performance could be unacceptable without powerful and low-latency inter-connection network. 

if the memory limitation  is caused by min-batch data, I think, data-parallelism could be enough cross more gpus, more machines. 

if the memory limitation is caused by high dimension sparse sample, please check sparse training in PaddlePaddle.

if the memory limitation is caused by some few long sequence sample with RNN, you can try to do specical processing. Some other system could use virtual memory to handle this special sample.

@yingfeng is there other requirement that drives the distributed model parall.?

**Comment by yingfeng** [comment]
@backyes I partly agree with what you have mentioned, although a real model parallelism is not practical currently due to both complexity and performance, actually, I think there are some cases where model parallelism is still useful:

- Model parallelism that could run on single machine to contain models on multiple GPU cards. Currently, Paddle still does not support such kinds of parallelism---the application context of ParallelNN is very limited.  In this case, there are several existing approaches might be feasible:
  - Assign whole layer to single device, and have the forward/backward run as a pipeline(something like the [LSTM example](https://github.com/dmlc/mxnet/blob/master/example/model-parallel-lstm/lstm.py) in MXNet). This could resolve part of the requirements that a single GPU device could not contain. Additionally, data parallelism should be necessary to multiple machines to improve the performance.
  - Split the layer into sub layers and have each of them run corresponding GPU device. This might be only suitable for fully connection layers, and such approach is only used within Amazon DSSTNE.
- Some solutions dedicated for CNN, such as "One Weired Trick", which decoulples conv layers from full connection layers and adopt model parallelism to the latter one. It seems baidu's "Deep Image" belongs to similar situation.

PS, although Tensorflow claims to support both data parallelism and model parallelism, it seems only similar approach as [LSTM example](https://github.com/dmlc/mxnet/blob/master/example/model-parallel-lstm/lstm.py) in MXNet is available, for other situations, I've not found any materials or code snippets to support.

**Comment by backyes** [comment]
Good comments ~ @yingfeng 
Firstly we have common sense with the data parallel. 

Then, 
We also have other options to address BIG model.
from the computer architecture view, memory limitation,  communication overhead within PCIe, even within nodes, can be unified. We can tradeoff the memory limitation with low latency inter-connection skill or latency hidden skills. Armed with this top view, using the High bandwidth link, low latency link, and latency hidden skill from observing that the time-sequential feature form forward and backward DEEP neural network could help to address BIG model issue.

For these cases that one BIG single layer or WIDDER layers instead of DEEP layers exists,  the method you mentioned could be preferred. 

Welcome more discussion.

**Comment by hedaoyuan** [comment]
Closed due to inactivity.
